Build failure: quickbms

### Steps To Reproduce

Steps to reproduce the behavior:
1. nix-shell -p quickbms

### Build log

https://gist.github.com/rvanlaar/decc031474bfb7fa67523e9e1cec27a0


### Notify maintainers

@samuelgrf 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
this path will be fetched (0.00 MiB download, 0.00 MiB unpacked):
  /nix/store/4l09rh46bvvp84d8xm879zrqkix4cyy1-nix-info
copying path '/nix/store/4l09rh46bvvp84d8xm879zrqkix4cyy1-nix-info' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 5.19.0-35-generic, Ubuntu, 22.04.2 LTS (Jammy Jellyfish), nobuild`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.2`
 - channels(roland): `"nixpkgs"`
 - nixpkgs: `/home/roland/.nix-defexpr/channels/nixpkgs`
```

It's possible that this package just isn't buildable
One of the errors encountered is:

```perform.c:1394:109: error: 'RSA_SSLV23_PADDING' undeclared (first use in this function); did you mean 'RSA_NO_PADDING'?```

`RSA_SSLV23_PADDING` is not supported in recent openssl versions.

--- a/pkgs/tools/archivers/quickbms/0002-disable-openssl.patch
+++ b/pkgs/tools/archivers/quickbms/0002-disable-openssl.patch
@@ -1,0 +1,10 @@
+--- src/Makefile	2023-05-25 18:16:56.000296653 +0900
++++ src/Makefile	2023-05-25 18:17:00.772272861 +0900
+@@ -23,7 +23,6 @@
+ EXTRA_TARGETS  = libs/amiga/amiga.s libs/powzix/*.cpp
+ CFLAGS += -msse2
+ endif
+-USE_OPENSSL	= 1
+ endif
+ 
+ # -liconv and -fPIC are necessary on Android

--- a/pkgs/tools/archivers/quickbms/default.nix
+++ b/pkgs/tools/archivers/quickbms/default.nix
@@ -1,4 +1,13 @@
-{ stdenv, lib, fetchzip, bzip2, lzo, openssl, zlib }:
+{ stdenv
+, lib
+, fetchzip
+, fetchpatch
+, bzip2
+, lzo
+, openssl_1_1
+, opensslSupport ? false
+, zlib
+}:
 
 stdenv.mkDerivation rec {
   version = "0.11.0";
@@ -9,7 +18,18 @@ stdenv.mkDerivation rec {
     hash = "sha256-uQKTE36pLO8uhrX794utqaDGUeyqRz6zLCQFA7DYkNc=";
   };
 
-  buildInputs = [ bzip2 lzo openssl zlib ];
+  patches = [
+    # Fix errors on x86_64 and _rotl definition
+    (fetchpatch {
+      name = "0001-fix-compile.patch";
+      url = "https://aur.archlinux.org/cgit/aur.git/plain/fix-compile.patch?h=quickbms&id=a2e3e4638295d7cfe39513bfef9447fb23154a6b";
+      hash = "sha256-49fT/L4BNzMYnq1SXhFMgSDLybLkz6KSbgKmUpZZu08=";
+      stripLen = 1;
+    })
+  ] ++ lib.optional (!opensslSupport) ./0002-disable-openssl.patch;
+
+  buildInputs = [ bzip2 lzo zlib ]
+    ++ lib.optional (opensslSupport) openssl_1_1;
 
   makeFlags = [ "PREFIX=$(out)" ];
 

